Add Ethereum provider type P3A metric

DJAndries opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves brave/brave-browser#25659

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

0. Start browser with fresh profile
1. Load any web page over https, check local state, ensure `Brave.Wallet.EthProvider` has a value of 2
2. Install and setup a competing wallet
3. Load any web page over https, ensure metric has a value of 1
4. Remove competing wallet
5. Change "default ethereum wallet" setting to "none"
6. Load any web page over https, ensure metric has a value of 0

darkdh commented on components/brave_wallet/browser/ethereum_provider_impl.cc:
```diff
@@ -11,6 +11,7 @@
 #include "base/containers/contains.h"
 #include "base/json/json_reader.h"
 #include "base/json/json_writer.h"
+#include "base/metrics/histogram_macros.h"
```
This is not needed.

darkdh commented on renderer/brave_wallet/brave_wallet_render_frame_observer.cc:
```diff
@@ -76,6 +78,41 @@ void BraveWalletRenderFrameObserver::DidCreateScriptContext(
+  v8::HandleScope handle_scope(isolate);
+  auto* web_frame = render_frame()->GetWebFrame();
+  v8::Local<v8::Context> context = web_frame->MainWorldScriptContext();
+  if (!EnsureConnected() || context.IsEmpty() || !CanCreateProvider()) {
```
we should not check `CanCreateProvider` here because not every third party wallets has the same security policy as us.

DJAndries replied:
Removing this is tricky. If we remove it (along with DOM feature policy check), then we may record `1` (aka third-party wallet) for the metric on a page that does not meet these conditions. Enforcing these conditions ensures that we give the native wallet a chance to load so that we can correctly evaluate the result.

DJAndries replied:
Perhaps the example I gave was not the best one. I'm still concerned about cases where the native wallet does not load due to the conditions defined in `CanCreateProvider` and `IsDOMFeaturePolicyEnabled`, but the third-party wallet does. Your original comment states that not every third-party wallet has the same security policy. Wouldn't that mean there are pages where a third-party wallet can load, but the native wallet cannot? These are the cases I'm concerned about.

darkdh replied:
>Wouldn't that mean there are pages where a third-party wallet can load, but the native wallet cannot?
Yes and that is a valid case and users can use third-party wallet on the page disregarding P3A report.
In that case, we should report third party wallet.

DJAndries replied:
Let's say the user had their default wallet setting to "Brave Wallet" (not preferring extensions). There would be some pages that do meet the mentioned conditions, where the native wallet would be the provider. The pages that do not meet the conditions will have a third-party wallet as the provider. This would mean you'd have some pages where `2` is recorded, and others where `1` is recorded. Is that correct?

darkdh replied:
That is right, when we can't provide our own `window.ethereum` even if default wallet setting is Brave Wallet.
In that case, 
if third party wallet insert `window.ethereum`, we report 1.
if no `window.ethereum` detected, we report 0

darkdh replied:
The whole point of this report is to tell whose `window.ethereum` users are using.

DJAndries replied:
This won't work for P3A, unfortunately. We can only report one answer per question/per week, so it's important that the reported answer is consistent. In the scenario that I described, it would seem that `2` is the preferred and strongest answer. It's clear that the user is using the native wallet as the preferred wallet. If we were to perform the changes that you suggested, a user may report `2` on one week and `1` on another. This is not desired from a P3A perspective.

darkdh replied:
Users can also change wallet setting multiple times in a week so the consistency claim here didn't convince me.
Without this change, there will be lots of 1 reported as 0 when default wallet setting is Prefer Extension. And that is the main reason I strongly recommend to remove those checks.

DJAndries replied:
If a user changes their wallet setting like you mentioned, and the reported P3A answer changes as a result, that is expected behavior. 

What isn't expected is having a volatile P3A answer for a user that never changes the setting, and uses the native wallet as a preferred wallet. It's probably safe to say that most users don't change this setting multiple times in a week. For these users, consistency absolutely matters since they will report the most answers for this question.

> Without this change, there will be lots of 1 reported as 0 when default wallet setting is Prefer Extension. 

I'm not sure when this will happen. I would imagine there would be some "1"s being reported as "2"s only if the third-party wallet had a more strict condition for loading their provider. Is it likely for a third-party wallet has a more strict condition? If so, then it could make sense to make the change. If it's likely the conditions are similar between our wallet and their wallet, then it makes no difference what we do. If it's less likely that the third-party wallet is more strict, then we should probably keep this current change.

DJAndries replied:
Whichever change produces the least amount of volatility is the best one, IMO

DJAndries replied:
Another thing to point out, is that `CanCreateProvider` checks for empty URLs, non-valid URLs and `about:blank` pages. If we were to remove this check, then we would probably get a "0" answer because neither the native wallet or third-party wallet would load.

darkdh replied:
> I'm not sure when this will happen assuming you mean "Brave Wallet (Prefer Extension)". I would imagine there would be some "1"s being reported as "2"s only if the third-party wallet had a more strict condition for loading their provider

Those won't be reported as 2, it will be never reported if those wallets have loosen policy. Just an easy example, if they allow wallet to be used on http page and we don't report it. We just do early return.

>Another thing to point out, is that CanCreateProvider checks for empty URLs, non-valid URLs and about:blank pages. If we were to remove this check, then we would probably get a "0" answer because neither the native wallet or third-party wallet would load.

I already said our wallet and extensions won't be loaded on those pages hence no `window.ethereum` will be created.

If you are so sticked with that edge case, where Brave Wallet is default wallet setting, and the page won't suffice our policy thus there will be only third party `window.ethereum` available for user to use, we can extra check wallet setting  in observer before reporting.

The point I'm trying to make is, when default setting is "Brave Wallet (Prefer Extension)", we should not add those policies on third party created `window.ethereum` because we have no control over what those extensions do.
If we do that and they failed to meet the condition then we won't get the "1" result.

DJAndries replied:
> Those won't be reported as 2, it will be never reported if those wallets have loosen policy. Just an easy example, if they allow wallet to be used on http page and we don't report it. We just do early return.

That's correct, the metric answer would not be reported for the http page with the way the PR works now. Accessing such pages would not result in reporting a "0" answer. Without the `CanCreateProvider` check, "1" would be reported which could be considered undesired, since it's likely that the user uses our native wallet for most purposes. 

The incorrect "2" answer I was referring to was due to a third party wallet that has a stricter policy, not a relaxed one. In this case, our native wallet would load, but the third party wallet would not, which would result in "2" being reported with the way the PR works now. Also not ideal, but at least we're only monitoring pages that have HTTPS and have valid DOM feature policies. I would imagine that most wallets (our wallet and competing wallets) would like these pages the most, so from my perspective it seems that the risk of a stricter third-party wallet policy breaking the code seems low.

> I already said our wallet and extensions won't be loaded on those pages hence no window.ethereum will be created.

Exactly, and without checking `CanCreateProvider`, we would report a "0" answer for such a page, which is undesired.

> If you are so sticked with that edge case, where Brave Wallet is default wallet setting, and the page won't suffice our policy thus there will be only third party window.ethereum available for user to use, we can extra check wallet setting in observer before reporting.

It could be argued that a third-party wallet having a stricter policy than our wallet is also an edge case, but you would know better than me (perhaps you can provide an example of such a wallet if so, just for my info?).

Checking the default wallet setting is a clever idea, and if it works, it could solve both of the cases that I mentioned above (this incorrect "1" answer case, and the incorrect "2" answer case). I will evaluate it.

darkdh commented on renderer/brave_wallet/brave_wallet_render_frame_observer.cc:
```diff
@@ -76,6 +78,41 @@ void BraveWalletRenderFrameObserver::DidCreateScriptContext(
+  if (!EnsureConnected() || context.IsEmpty() || !CanCreateProvider()) {
+    return;
+  }
+  if (render_frame()->GetWebFrame()->GetDocument().IsDOMFeaturePolicyEnabled(
+          context, "ethereum")) {
```
same reason as above, we don't need this policy when evaluating `window.ethereum`, otherwise the P3A result will be inaccurate for third party wallets. Those policies should only affect whether we should create `window.ethereum` or not.

DJAndries replied:
see comment above

bbondy replied:
Why won't we:
- For other wallets: Allow P3A to report for 3p wallets whether or not `CanCreateProvider`
- For our wallet: In the case that we can create the provider we return a different value than when we can't.

bbondy replied:
> For our wallet: In the case that we can create the provider we return a different value than when we can't.

I.e. we report either Brave Wallet or None. But I agree with @darkdh about we should be reporting about the 3p wallet whether or not `CanCreateProvider`.

DJAndries replied:
@bbondy When you say "a different value", do you mean another P3A answer?

DJAndries replied:
> I.e. we report either Brave Wallet or None. 

If the provider was created, I believe the code would have to choose between "third-party" and "Brave Wallet". The only answer we could rule out is "none". If I'm missing something, feel free to correct

bridiver replied:
I think this is unnecessarily complicating things. All we need to do is check to see if `window.ethereum` exists and if it does, check for `window.ethereum.isBraveWallet`. No other checks are needed here

bridiver replied:
also can we move this out of BraveWalletRenderFrameObserver and put it in BraveWalletP3ARenderFrameObserver or something? Or at the very least add a util method that encapsulates all of this? We should avoid polluting code with P3A specifics where possible

DJAndries replied:
> I think this is unnecessarily complicating things. All we need to do is check to see if window.ethereum exists and if it does, check for window.ethereum.isBraveWallet. No other checks are needed here

@bridiver please see thread above regarding unintended consequences. We will need to perform at least one additional check. I'm evaluating an alternate solution at the moment

bridiver replied:
so after talking to @DJAndries I think there is some general confusion here that would be cleared up with some comments

darkdh commented on renderer/test/js_ethereum_provider_browsertest.cc:
```diff
@@ -123,10 +127,16 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, AttachOnReload) {
   ReloadAndWaitForLoadStop();
+
+  histogram_tester_->ExpectBucketCount("Brave.Wallet.EthProvider", 2, 1);
+
```
we should add `histogram_tester_->ExpectBucketCount("Brave.Wallet.EthProvider", 1, 1);`
after
https://github.com/brave/brave-core/blob/bbcc780dfb285b07b83f5663875cac343fbc94fa/renderer/test/js_ethereum_provider_browsertest.cc#L152-L154

darkdh replied:
I think that is ok if we leave this out.

bridiver commented on renderer/brave_wallet/brave_wallet_render_frame_observer.cc:
```diff
@@ -76,6 +78,42 @@ void BraveWalletRenderFrameObserver::DidCreateScriptContext(
   }
 }
 
+void BraveWalletRenderFrameObserver::DidFinishLoad() {
```
why are you doing this in DidFinishLoad?

DJAndries replied:
I assumed that the page load was a reasonable deadline for a third-party extension to inject their provider

bridiver commented on components/brave_wallet/browser/brave_wallet_service.h:
```diff
@@ -223,6 +223,9 @@ class BraveWalletService : public KeyedService,
 
   void OnOnboardingShown() override;
 
+  void ReportEthereumProviderForP3A(
```
why are we adding this to BraveWalletService? We should create a mojo interface specifically for brave_wallet_p3a

DJAndries replied:
I agree that a separate interface for P3A is ideal. Looks like it'll require a significant refactor, since the wallet service owns the p3a object

bridiver commented on renderer/brave_wallet/brave_wallet_render_frame_observer.cc:
```diff
@@ -76,6 +78,42 @@ void BraveWalletRenderFrameObserver::DidCreateScriptContext(
 }
 
+void BraveWalletRenderFrameObserver::DidFinishLoad() {
+#if !BUILDFLAG(IS_ANDROID)
```
why is this buildflag here?

DJAndries replied:
We only want to report this question for desktop

DJAndries replied:
will add comment

DJAndries commented:
Refactored. Changes made:

- Extracted essential checks from `CanCreateProvider` into `IsPageValid` so we can avoid checking the ethereum provider on non http/https pages, such as `about:blank` or `brave://`
- Created `BraveWalletRenderFrameObserverP3AUtil` to handle ethereum provider evaluation
- Check default ethereum provider setting during evaluation to handle edge cases

iefremov reviewed (approved):
unlocking components/p3a

darkdh reviewed (approved):
lgtm now

bridiver commented on browser/brave_content_browser_client.cc:
```diff
@@ -270,6 +272,23 @@ void BindCosmeticFiltersResources(
+    mojo::PendingReceiver<brave_wallet::mojom::BraveWalletP3A> receiver) {
+  auto* context = frame_host->GetBrowserContext();
+  if (brave_wallet::IsAllowedForContext(frame_host->GetBrowserContext())) {
+    brave_wallet::BraveWalletService* wallet_service =
```
this is a bit messy to get the brave wallet p3a from BraveWalletService and it also creates a bit of a mess for unit testing https://github.com/brave/brave-core/pull/20433/files#diff-624bd28f2b67dac7bfc8188ddf0b6df7e5a43e83f7820d827df57564024fff1eR114
It doesn't seem like BraveWalletService needs to have a reference to BraveWalletP3A at all because it's only used for `OnWalletUnlockPreferenceChanged` which `BraveWalletP3A` could do directly. Seems better to have it as a separate keyed service factory which is null for testing
